src/validators/model_fields.rs

```diff
@@ -372,8 +372,24 @@ impl Validator for ModelFieldsValidator {
             }
         }?;
 
+        let new_extra = match &self.extra_behavior {
+            ExtraBehavior::Allow => {
+                let non_extra_data = PyDict::new(py);
+                self.fields.iter().for_each(|f| {
+                    let popped_value = PyAny::get_item(new_data, &f.name).unwrap();
+                    new_data.del_item(&f.name).unwrap();
+                    non_extra_data.set_item(&f.name, popped_value).unwrap();
+                });
+                let new_extra = new_data.copy()?;
+                new_data.clear();
+                new_data.update(non_extra_data.as_mapping())?;
+                new_extra.to_object(py)
+            }
+            _ => py.None(),
+        };
```

Comment:
I feel there must be a better way to achieve this — @davidhewitt maybe you have some suggestions?

Reply:
This looks pretty sound to me on the whole

Reply:
You're modifying `new_data` so I don't think you need to create two new dicts. The challenge seems to be that you only have the list of known fields, so you are forced to remove them from `new_data`.

How about swapping the binding over like this:

```suggestion
        let (new_data, new_extra) = match &self.extra_behavior {
            ExtraBehavior::Allow => {
                // Move non-extra keys out of new_data, leaving just the extra in new_data
                let non_extra_data = PyDict::new(py);
                for field in &self.fields
                    let popped_value = PyAny::get_item(new_data, &field.name).unwrap();
                    new_data.del_item(&f.name).unwrap();
                    non_extra_data.set_item(&f.name, popped_value).unwrap();
                }
                (non_extra_data, new_data.to_object())
            }
            // FIXME do you need to throw if `new_data` contains any extra keys?
            _ => (new_data, py.None()),
        };
```

Reply:
I think we check previously for any possible source of extra keys (and error), so I think we can ignore that potential issue here.

Reply:
just kidding, this breaks some tests because it assumes in some places that `validate_assignment` does an in-place modification to the fields dict, but this makes it become the `extra` dict. I'm not sure what the ideal behavior is here but I am inclined to leave as it was before for now, and we can revisit this in a separate PR if/when it seems worthwhile